### ISSUE TITLE
Reading from stdin on Windows is unworkable

### ISSUE THREAD

**Issue by Tetralux** [odin-lang/Odin#1214]
## Context

Using the core library, there is currently no straightforward way to read a line of input from stdin on Windows.

This is because `os.read` is actually a form of "read-full"; that is to say that it reads repeatedly until the buffer that was provided by the user is filled, or the OS reports an error.

On Linux, this is not the case because `os.read` only calls the read syscall, which just does a short-read, and that's the end of it.

A while ago, I wrote a paste which implements a `read_until` procedure which reads from a handle until a specified delimeter byte; either reading into a user-provided slice, or by allocating a dynamic array until the delimeter is found.

It addresses this problem by including a `read_some` procedure; a procedure which is able to do a short-read, and works on all platforms: https://p.teknik.io/Raw/IT996

`core:io` and `core:bufio` were meant to address this in the core library, but they are unable -- while they do provide a mechanism for reading up-to-and-including a delimeter, they still fall flat because of the long-standing problematic nature of `os.read` on Windows.

## Reproduction

```odin
package main

import "core:fmt"
import "core:os"
import "core:bufio"
import "core:io"
import "core:strings"

main :: proc() {
	s := os.stream_from_handle(os.stdin)
	r: bufio.Reader
	bufio.reader_init(&r, io.Reader{s})

	line, err := bufio.reader_read_string(&r, '\n')
	assert(err == .None)

	trimmed := strings.trim_space(line)

	fmt.printf("read:    %q\n", line)
	fmt.printf("trimmed: %q\n", trimmed)
}
```
Note that if you want to avoid allocation, you'd have to use `reader_read_slice` instead, which uses the `Reader`'s buffer, but gets you a `[]byte` instead of a `string`, so an extra cast is needed:
```odin
package main

import "core:fmt"
import "core:os"
import "core:bufio"
import "core:io"
import "core:strings"

main :: proc() {
	s := os.stream_from_handle(os.stdin)
	r: bufio.Reader
	bufio.reader_init(&r, io.Reader{s})

	bytes, err := bufio.reader_read_slice(&r, '\n') // add reader_read_slice_string() ?
	assert(err == .None)

	line := string(bytes)
	trimmed := strings.trim_space(line)

	fmt.printf("read:    %q\n", line)
	fmt.printf("trimmed: %q\n", trimmed)
}
```

**Comment by gingerBill** [comment]
The main reason for this is that on Windows, if a file is a console, you cannot use `ReadFile` on it but require `ReadConsoleW` + UTF1-6 to UTF-8 conversion.

This means that we probably need to check whether the file is a console or not before calling the internal Windows calls.

**Comment by Tetralux** [comment]
Fixed in 6616882708f9ad526d215cbe97b1a3f12dfd6bf8.
